### ISSUE TITLE
wheelinstaller does not use the archive when installing sdist

### ISSUE THREAD

**Issue by chrisirhc** [bazelbuild/rules_python#2090]
# 🐞 bug report

Version used: https://github.com/bazelbuild/rules_python/commit/59bb4a88783cca738394cb3cb91bc8243df5e18d
When using `experimental_index_url`, the index url is not respected when falling back to sdist, it is also not using the downloaded archive.

### Affected Rule

```

INFO: Repository rules_python~~pip~third_party_python_piper_39_caesar_sdist_37026b9b instantiated at:
--
<builtin>: in <toplevel>
Repository rule whl_library defined at:
/var/cache/bazel/output_base/external/rules_python~/python/private/pypi/whl_library.bzl:451:30: in <toplevel>
ERROR: An error occurred during the fetch of repository 'rules_python~~pip~third_party_python_piper_39_caesar_sdist_37026b9b':
Traceback (most recent call last):
File "/var/cache/bazel/output_base/external/rules_python~/python/private/pypi/whl_library.bzl", line 243, column 35, in _whl_library_impl
repo_utils.execute_checked(
File "/var/cache/bazel/output_base/external/rules_python~/python/private/repo_utils.bzl", line 204, column 29, in _execute_checked
return _execute_internal(fail_on_error = True, *args, **kwargs)
File "/var/cache/bazel/output_base/external/rules_python~/python/private/repo_utils.bzl", line 135, column 20, in _execute_internal
logger.fail((
File "/var/cache/bazel/output_base/external/rules_python~/python/private/repo_utils.bzl", line 85, column 39, in lambda
fail = lambda message_cb: _log(-1, "FAIL", message_cb, fail),
File "/var/cache/bazel/output_base/external/rules_python~/python/private/repo_utils.bzl", line 75, column 16, in _log
printer("\nrules_python:{} {}:".format(
…
rules_python:whl_library(@@rules_python~~pip~third_party_python_piper_39_caesar_sdist_37026b9b) FAIL: repo.execute: whl_library.BuildWheelFromSource(rules_python~~pip~third_party_python_piper_39_caesar_sdist_37026b9b, caesar==7.0.0): end: failure:
--
command: /var/cache/bazel/output_base/external/rules_python~~python~python_3_9_host/python -m python.private.pypi.whl_installer.wheel_installer --requirement caesar==7.0.0 --isolated --extra_pip_args "{\"arg\":[]}" --pip_data_exclude "{\"arg\":[]}" --environment "{\"arg\":{}}"
return code: 1
…
  | ERROR: Could not find a version that satisfies the requirement caesar==7.0.0 (from versions: 0.1.2, 0.1.3)
-- | --
ERROR: No matching distribution found for caesar==7.0.0
…
subprocess.CalledProcessError: Command '['/var/cache/bazel/output_base/external/rules_python~~python~python_3_9_host/python', '-m', 'pip', '--isolated', 'wheel', '--no-deps', '-r', '/tmp/tmpzb5a5smh']' returned non-zero exit status 1.

```

 The issue is caused by the rule: 
The likely cause is in https://github.com/bazelbuild/rules_python/blob/59bb4a88783cca738394cb3cb91bc8243df5e18d/python/private/pypi/whl_library.bzl#L233

### Is this a regression?

Not sure.

### Description

 A clear and concise description of the problem...

## 🔥 Exception or Error

<pre><code>

</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>

  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>

  </code>
</pre>

**Rules_python version:**

<pre>
  <code>

  </code>
</pre>

**Anything else relevant?**

**Comment by chrisirhc** [comment]
Found a potential fix #2091 , but hit a new issue:

```

  | ===== stdout end =====
-- | --
===== stderr start =====
error: subprocess-exited-with-error
× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> [3 lines of output]
Looking in links: .
ERROR: Could not find a version that satisfies the requirement setuptools>=40.8.0 (from versions: none)
ERROR: No matching distribution found for setuptools>=40.8.0
[end of output]
note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error
× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.
note: This error originates from a subprocess, and is likely not a problem with pip.
Traceback (most recent call last):
```

**Comment by aignas** [comment]
Could you please create a minimum reproducible example?
